vignettes/rspm-snaphots.Rmd

```diff
@@ -0,0 +1,223 @@
+If you have good arguments why another snapshot might be helpful for a particular R version, you are welcome to open PR and share your thoughts with us!
+
+Assuming that you have updated/downgraded your snapshot, the last step is to update the actual installed packages.
+Now that there is another snapshot in place, you have new package sources available and should synchronize your project library to it.
+
+That said, for updating call `renv::update()`.
+{renv} will show you a list of changes which will be applied and which packages will be updated to which version.
+Here a short example:
+
+    - gh             [1.2.0 -> 1.3.0]
+    - highr          [0.8 -> 0.9]
+    - jquerylib      [0.1.3 -> 0.1.4]
+
+Last, call `renv::snapshot()` to record the new versions in the lockfile.
+
+### Downgrading snapshots and keeping the same R version
+
+Is a bit more complicated.
+There is no dedicated `downgrade()` function because by default this function would not know which version to match against for the downgrade (remember, RSPM snapshots are not used by default by {renv}).
+One option is to go with `renv::revert()` and restore the lockfile contents of a previous commit.
+However, this approach has the downside that it does not account for possible new changes to the lockfile one has made meanwhile.
+
+What you actually want is to restore all packages listed in the lockfile with their version available in the configured snapshot.
+
+``` {.r}
+installed_pkgs = unname(installed.packages(lib.loc = .libPaths()[1])[, "Package"])
+renv::install(installed_pkgs)
+```
+
+This will install all installed packages again using the version of the snapshot date listed in `renv.lock`.
+Luckily, there is a function in {cynkrathis} to simplify this: `renv_downgrade()`.
+It will restart the session to ensure it is picking up the correct repo option and then execute the calls shown above.
+
+Next, call `renv::snapshot()` and you're good to go.
+
+The output should look similar as the following one
+
+    - testthat       [3.0.2 -> 3.0.1]
+    - textshaping    [0.3.4 -> 0.2.1]
+
+## Default {renv} options and settings
+
+We also recommend to set some default {renv} options and settings to make project work even more enjoyable.
+There are more potential settings which might need adjustments when administrating a RStudio Workbench instance for multiple people.
+
++-------------------------------------+-----------------+-----------------+
+| Name                                | Value           | Type            |
++=====================================+=================+=================+
+| `renv.config.auto.snapshot`         | FALSE           | option          |
++-------------------------------------+-----------------+-----------------+
+| `RENV_PATHS_PREFIX_AUTO`            | TRUE            | Env var         |
++-------------------------------------+-----------------+-----------------+
+
+- `renv.config.auto.snapshot`: @kirill FIXME
```

Comment:
Keep FALSE.

Reply:
I would have liked you to add some more details to why auto-snapshotting might raise issues in certain situations to explain why we suggest to turn it off by default.

Reply:
I can write a one-paragraph Q&A. The default is off, do we even need to explain it?

Reply:
I think this is a special case since I see many people turning this on and liking it (and probably never hitting the edge cases) so it could be worth explaining some of these here.

Reply:
> many people (citation needed)

Okay, here we go.

I have auto-snapshotting on. I pull from GitHub, `renv.lock` changes.

1. For some reason I forget to restart the R session and don't get a reminder that my package library is out of sync. I install a new package, auto-snapshot kicks in, the `renv.lock` is a mess made from a combination of my outdated library and the new packages. I trust a lot in automation and commit, somebody else sorts this out.
2. Worse, I do get a reminder, I remember to restore, and then the restore fails. I frantically try to install packages manually, it works, but then auto-snapshot writes a bogus state into the lock file.
3. If everything works out great, and I actually manage to restore cleanly and install new packages, now everybody immediately sees all my intermediate experiments with packages that I install to test-drive and then later remove. "Ideally", I commit straight to the main branch, so that everybody can track the flow of my exploration and happily install all the packages I've touched to their library, to later remove them.

These are just a few scenarios that come to my mind and actually happened in practice. With manual snapshots, we decouple snapshotting from restoring and installing, which gives us a bit more control. This works because in a multi-user project there is often one or few contributors who define the package environment, and the others just follow (and don't need this setting to begin with).

Maybe the following narrative could be a compromise: users who actively manage the project's package library and carefully review what they commit to the lock file are allowed to turn this setting on if it helps them, but the default (off) ultimately gives more control. Updates to the lock file should not be done willy-nilly, either way, for the update of each package there should be a reason -- especially in larger projects. This is why I feel uncomfortable recommending an overall lifting for each package update.

Reply:
Good arguments, I agree. Thanks for outlining! This could be very valuable for a lot of people.